interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,319 @@
+  private int bandwidth;
+  private long latency;
+  private final BlockingQueue<Message> serverQueue = new DelayQueue<Message>();
+  private final BlockingQueue<Message> clientQueue = new DelayQueue<Message>();
+  private volatile int serverQueueSize;
+  private volatile int clientQueueSize;
+  private volatile boolean shutDown;
+  private ServerSocket clientAcceptor;
+  private Socket serverSock;
+  private Socket clientSock;
+  private List<Thread> workers = new ArrayList<Thread>();
+
+  /**
+   * Returns a new TrafficControlProxy with default bandwidth and latency
+   */
+  public TrafficControlProxy() {
+    this(DEFAULT_BAND_BPS, DEFAULT_DELAY_NANOS, TimeUnit.NANOSECONDS);
+  }
+
+  /**
+   * Returns a new TrafficControlProxy with bandwidth set to targetBPS, and latency set to
+   * targetLatency in latencyUnits.
+   *
+   * @param targetBPS
+   * @param targetLatency
+   * @param latencyUnits
+   */
+  public TrafficControlProxy(int targetBPS, int targetLatency, TimeUnit latencyUnits) {
+    checkArgument(targetBPS > 0);
+    checkArgument(targetLatency > 0);
+    bandwidth = targetBPS;
+    // divide by 2 because latency is applied in both directions
+    latency = latencyUnits.toNanos(targetLatency) / 2;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength / 4);
+  }
+
+  /**
+   * Convenience method to change bandwidth/latency without restarting.
+   *
+   * @param targetBPS
+   * @param targetLatency
+   * @param latencyUnits
+   * @throws IOException
+   */
+  public void reset(int targetBPS, int targetLatency, TimeUnit latencyUnits) throws IOException {
+    checkArgument(targetBPS > 0);
+    checkArgument(targetLatency > 0);
+    shutDown();
+    bandwidth = targetBPS;
+    queueLength = (int) Math.max(bandwidth * latency / TimeUnit.SECONDS.toNanos(1), 1);
+    chunkSize = Math.max(1, queueLength / 4);
+    serverQueueSize = 0;
+    clientQueueSize = 0;
+    serverQueue.clear();
+    clientQueue.clear();
+  }
+
+  /**
+   * Waits for client and server and start reader/writer threads.
+   *
+   * @throws IOException
+   */
+  public void start() throws IOException {
+    // ClientAcceptor uses a ServerSocket server so that the client can connect to the proxy as it
+    // normally would a server. serverSock then connects the server using a regular Socket as a
+    // client normally would.
+    logger.info("Starting new proxy on port " + proxyPort);
+    clientAcceptor = new ServerSocket();
+    clientAcceptor.setReuseAddress(true);
+    clientAcceptor.bind(new InetSocketAddress(loopback, proxyPort));
+    clientSock = clientAcceptor.accept();
+    serverSock = new Socket();
+    serverSock.connect(new InetSocketAddress(loopback, serverPort));
+
+    startWorkers();
+  }
+
+  public void shutDown() throws IOException {
+    logger.info("Proxy shutting down... ");
+    shutDown = true;
+    clientAcceptor.close();
+    clientSock.close();
```

Comment:
nit: if one of these fails, the subsequent closes won't work.

Maybe add a TODO to hand this case?